Fixed #27688 -- Made messages' add_message() request check use ducktyping.

rsalmaso opened the pull request:
ticket https://code.djangoproject.com/ticket/27688

timgraham commented on tests/messages_tests/test_api.py:
```diff
@@ -45,10 +48,52 @@ def test_middleware_missing(self):
+class CustomMiddlewareApiTest(ApiTest):
+    @modify_settings(MIDDLEWARE={'remove': ['messages_test.api.CustomMessageMiddleware']})
+    def test_middleware_missing(self):
+        print("custom")
```
debug

timgraham commented on django/contrib/messages/api.py:
```diff
@@ -18,16 +18,18 @@ def add_message(request, level, message, extra_tags='', fail_silently=False):
+                'django.contrib.messages.middleware.MessageMiddleware'
+            )
+        else:
+            raise TypeError("add_message() argument must be an HttpRequest like object, "
```
It's not clear to me why it's valuable to change the semantics here so that this validation doesn't happen first if `fail_silently=False`.
I don't think the message must be changed to add "like".

rsalmaso replied:
1) patch updated
2) ok, I was thinking about rest_framework.request.Request which is not HttpRequest. i'll revert to original string

timgraham commented on django/contrib/messages/api.py:
```diff
@@ -18,16 +18,18 @@ def add_message(request, level, message, extra_tags='', fail_silently=False):
-            'You cannot add messages without installing '
-            'django.contrib.messages.middleware.MessageMiddleware'
-        )
+        if request and "django.contrib.messages.middleware.MessageMiddleware" not in settings.MIDDLEWARE:
```
I'm uncertain why you want to prohibit middleware subclasses?

rsalmaso replied:
In reality I was trying to avoid that (with a poorly attempt). Anyway patch rewritten.

rsalmaso commented:
Patch rewritten and rebased on https://github.com/django/django/commit/07931826947bf68925cba9ca89b95636f9609fbd
It use the "Ask forgiveness, not permission" pattern instead of using hasattr().
The middleware check is gone, and so no need to have a test for it.
I've added an explicit commit for removing the MessageMiddleware from settings.

timgraham commented:
I'm unsure why the second commit is needed.

rsalmaso commented:
Just for clarity, as I think they are two different steps. But I can squash and have only one commit if preferred.

timgraham commented:
What I meant is that I don't see any test failures after the first commit so I don't see why the second commit is required.

rsalmaso commented:
Oh ok, now I understand.
Currently `test_middleware_missing(_silently)` should test whenever `MessageMiddleware` is missing/not installed. But what they really do is mocking MessageMiddleware calls, leaving current settings unmodifed.
I find this a clearer intent about that the setting is gone.
I've removed the commit, if needed I'll add another ticket/pr for it.

timgraham commented:
I force pushed some edits to your branch, however, the test doesn't act as a regression test (it's passing on master already). I think `SimpleLazyObject` proxies the `isinstance` call, so you'll need a different approach.

rsalmaso commented:
Yes, `SimpleLazyObject` is really smart about `isinstance`.
I used the django-rest-framework version as test case, so can acts as regression for previus `add_message()` implementation.

timgraham commented:
Is that the simplest version of `CustomRequest` that could be used? The exception handling seems complicated to add for this relatively simple change.

rsalmaso commented:
I simplyfied the CustomRequest as much as possible (really, the full drf version is not useful here).
